docs/okeps/okep-4380-network-qos.md

```diff
@@ -0,0 +1,555 @@
+packet’s IP header. So, the marking appears inside the GENEVE tunnel. There are ways to transfer
+this marking to outer header and influence how the underlay network fabric should handle such
+packets, however that is outside the scope of this proposal.
+
+Kubernetes offers partial support for QoS features through annotations such
+as `kubernetes.io/egress-bandwidth` and `kubernetes.io/ingress-bandwidth` at the Pod interface
+level. However, these annotations lack fine-grained control, as they cannot target specific types of
+traffic (e.g., video streaming) on an interface. The Network Plumbing Working Group (NPWG) has
+extended these annotations to secondary networks, but they remain limited to interface-level
+configurations without options for selecting a particular traffic flow. Additionally, Kubernetes
+currently lacks an API for DSCP packet marking.
+
+To address these limitations, this proposal introduces a NetworkQoS API that enables fine-grained
+bandwidth enforcement and packet marking across all interfaces within a Pod.
+
+## User-Stories/Use-Cases
+
+#### Story 1
+
+```text
++--------+                                                                                   
+|NS1/Pod1+--DSCP:20                                                                          
++--------+     |         .-----.----------------------------.-----.                          
++--------+     +------> ;       :     Overlay Traffic      ;       :   +--------------------+
+|NS1/Pod2+--DSCP:11---->:       ;  Various DSCP markings   :       --->|NS3/ReverseProxy Pod|
++--------+      +------> \     /                            \     /    +--------------------+
+                |         `---'------------------------------`---'                           
++--------+      |                  .-------------------.                                     
+|NS2/Pod3+--DSCP:30          _.---'                     `----.                               
++--------+                  /        Physical Underlay        \                              
+                           (     (unaware of DSCP marking)     )                             
+                              `----.                   _.---'                                
+                                    `-----------------'                                      
+```
+
+As an OVN-Kubernetes user, I want to configure DSCP marking for egress east-west overlay traffic
+
+In the diagram above, I want Pods Pod1 and Pod2 in Namespace NS1 to be treated differently by
+the ReverseProxy Pod. The packets leaving Pods Pod1 and Pod2 will be marked with DSCP value of
+20 and 11 respectively. These marking will be retained on the overlay across the fabric and
+arrive at the RP pod where the packets can be treated differently based on these markings,
+offering tailored processing or forwarding behavior.
+
+The application owners have the flexibility to define how they utilize the 6-bit DSCP field to
```

Comment:
Can you use an example of some specific applications for pod1 and pod2 to illustrate the actual use case?